nx/lib/nx/binary_backend.ex

```diff
@@ -1571,6 +1571,65 @@ defmodule Nx.BinaryBackend do
     end
   end
 
+  @impl true
+  def scatter_add(
+        %T{} = out,
+        %T{shape: shape} = target,
+        %T{shape: {indices_rows, _indices_cols} = indices_shape} = indices,
+        %T{shape: {indices_rows}} = updates,
+        _opts
+      ) do
+    indices_bin_list =
+      indices |> to_binary() |> aggregate_axes([1], indices_shape, elem(indices.type, 1))
+
+    indices_list =
+      match_types [indices.type] do
+        for idx <- indices_bin_list do
+          {_, reversed} =
+            for <<match!(x, 0) <- idx>>, reduce: {Tuple.to_list(shape), []} do
+              {[current_dim | tl], acc} ->
+                item = read!(x, 0)
+
+                unless item < current_dim do
+                  # TO-DO: better error message
+                  raise ArgumentError, "invalid index received for tensor"
+                end
+
+                {tl, [item | acc]}
+            end
+
+          Enum.reverse(reversed)
+        end
+      end
+
+    updates_list =
+      match_types [updates.type] do
+        for <<match!(x, 0) <- to_binary(updates)>>, do: read!(x, 0)
+      end
+
+    indices_with_updates =
+      indices_list
+      |> Enum.zip(updates_list)
+      |> Enum.group_by(fn {idx, _} -> idx end, fn {_, upd} -> upd end)
+      |> Enum.map(fn {idx, upds} -> {idx, Enum.sum(upds)} end)
+
+    for {idx, upd} <- indices_with_updates, reduce: target do
+      t ->
+        put_slice(
+          out,
+          t,
+          idx,
+          Nx.tensor(upd) |> Nx.reshape(Tuple.duplicate(1, length(idx))),
+          fn x, y ->
+            x = from_binary(%T{names: [], shape: {}, type: t.type}, x)
+            y = from_binary(%T{names: [], shape: {}, type: updates.type}, y)
+
+            Nx.add(x, y) |> Nx.as_type(out.type) |> Nx.to_binary()
+          end
+        )
+    end
```

Comment:
@seanmor5 @josevalim @jonatanklosko does anybody know of a better way to do this "traverse and update" for the tensor?

I thought of recursevely using aggregate_axes until the last dim is reached, having the accumulated indices so far, but perhaps there's a better solution.

Although this implementation works, I think we can do better.

Reply:
Nevermind! Looking at gather code I had an idea which might work :)
https://github.com/elixir-nx/nx/pull/456/files#diff-15095458a1b2fb983edb557f9d4b0664e25425d06d053665f48867c8df015307R1784-R1804

I plan on producing a list of {offset, increment} for each of the index-update pairs, and then just travelling across the binary jumping between offsets. This can be done in O(indices) instead of O(binary_elements), I think.

Thanks for the inspiration @jonatanklosko

Reply:
Yup, I think you can iteratively use `binary_part` for the unchanged areas and put the updates inbetween, then just binary-concat everything. That's probably what you meant :D

Reply:
Yeah! There'll be some type conversion needed along the way, but that's the idea :D